lib/core/utils/node-serializer.js

```diff
@@ -0,0 +1,40 @@
+import assert from './assert';
+import DqElement from './dq-element';
+
+let serializer = null;
+
+/**
+ * @param {Object} newSerializer
+ * @property {Function} toSpec Converts an Element or VirtualNode to a "spec", a form
+ * suitable for JSON.stringify to consume. Output must include all properties
+ * that DqElement.toJSON() would have. Will always be invoked from the
+ * input element's original page context.
+ * @property {Function} mergeSpecs Merges two specs (produced by toSpec) which
+ * represent element's parent frame and an element, respectively. Will
+ * *not* necessarily be invoked from *either* node's original page context.
+ * This operation must be associative, that is, these two expressions must
+ * produce the same result:
+ * - mergeSpecs(a, mergeSpecs(b, c))
+ * - mergeSpecs(mergeSpecs(a, b), c)
+ */
+export default function nodeSerializer(newSerializer) {
+  assert(typeof serializer === 'object', 'serializer must be an object');
+  serializer = newSerializer;
+}
+
+Object.assign(nodeSerializer, {
+  toSpec(node) {
+    const dqElm = new DqElement(node);
+    if (typeof serializer?.toSpec === 'function') {
+      return serializer.toSpec(dqElm);
+    }
+    return dqElm.toJSON();
```

Comment:
Lets discuss this when @dbjorge is back.

Reply:
Discussed with @dbjorge and he doesn't see any problems with the performance shift, and even thinks that this shift better reflects the time required in to run a rule from the standpoint of the user. So we're OK with this change.

My only request is that in a separate PR (I'll create a tech debt ticket) we log the time for serializing a node so that we can make sure we can separate it from the performance metric for running a rule and check.

Reply:
[API documentation](https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#use-resulttypes) recommends using resultTypes to improve performance. With this change, does that recommendation become obsolete, because selectors are now always generated before we even know if it's a violation or not? I have this same concern with switching from run to runPartial in our codebase, but I haven't measured the impact yet, so not sure if it's a big problem in practice. There might also be users that intentionally set selectors: false for performance reasons and use ancestry or something else instead. I guess once a custom node serializer can be provided, users could use that + custom reporter to fully customize this behavior.

Reply:
@hamax That's a good question, and yes, we might change that with #4124. The reason I did it this way is because some of these properties are used before they are filtered. `nodeIndexes` is actually never reported on, it is exclusively used in `after` methods. Ancestry is also used in `after` methods.

If we apply those filters too soon we lose data that's still needed. We can get smarter about what is safe to filter early and what isn't, but that's out of scope for this PR.